src/algebra/group/hom.lean

```diff
@@ -179,7 +179,45 @@ omit mP
 @[simp, to_additive] lemma comp_id (f : M →* N) : f.comp (id M) = f := ext $ λ x, rfl
 @[simp, to_additive] lemma id_comp (f : M →* N) : (id N).comp f = f := ext $ λ x, rfl
 
-variables [mM] [mN]
+end monoid_hom
+
+section End
+
+variables (M) [monoid M]
+variables (A : Type*) [add_monoid A]
+
+/-- The monoid of endomorphisms. -/
+protected def monoid.End := M →* M
+
+instance monoid.End.monoid : monoid (monoid.End M) :=
+{ mul := monoid_hom.comp,
+  one := monoid_hom.id M,
+  mul_assoc := λ _ _ _, monoid_hom.comp_assoc _ _ _,
+  mul_one := monoid_hom.comp_id,
+  one_mul := monoid_hom.id_comp }
+
+instance monoid.End.inhabited : inhabited (monoid.End M) :=
+⟨1⟩
```

Comment:
Could you please define `has_coe_to_fun` and add lemmas `coe_one : ((1 : monoid.End M) : M → M) = id` and `coe_mul : ((f * g : monoid.End M) : M → M) = f ∘ g` in `monoid` and `add_monoid` namespaces?

Reply:
done